ivadomed/inference.py

```diff
@@ -299,9 +319,9 @@ def segment_volume(folder_model: str, fname_images: list, gpu_id: int = 0, optio
         fname_images (list): list of image filenames (e.g. .nii.gz) to segment. Multichannel models require multiple
             images to segment, e.i., len(fname_images) > 1.
         gpu_id (int): Number representing gpu number if available. Currently does NOT support multiple GPU segmentation.
-        options (dict): Contains postprocessing steps and prior filename (fname_prior) which is an image filename
-            (e.g., .nii.gz) containing processing information (e.i., spinal cord segmentation, spinal location or MS
-            lesion classification)
+        options (dict): Contains postprocessing steps, film metadata, microscopy PixelSize and overlap_2D metadata, and
+            prior filename (fname_prior) which is an image filename (e.g., .nii.gz) containing processing information
+            (e.g., spinal cord segmentation, spinal location or MS lesion classification).
             e.g., spinal cord centerline, used to crop the image prior to segment it if provided.
             The segmentation is not performed on the slices that are empty in this image.
```

Comment:
I did a quick search for `options (dict)` and it looks like we use the same argument in multiple functions in `inference.py`.

Do we need to keep this description consistent across all of the functions?

Reply:
Also, this description is starting to feel a little unclear to me...

We have a text description (`"Contains postprocessing steps, film metadata, microscopy PixelSize and overlap_2D metadata, and prior filename (fname_prior) which is an image filename"`), but it doesn't tell me explicitly which dict keys are supported (or required, even). (`PixelSize`, `fill_holes`, `keep_largest`, etc.)

I'm wondering if we could rewrite the text description into a list (just a rough sketch, probably inaccurate):

```python
        options (dict): This can optionally contain any of the following key-value pairs:
            * 'fill_holes': (bool) <description for this postprocessing step>
            * 'keep_largest': (bool) <and so on, etc.>
            * 'overlap_2D': (bool)
            * 'PixelSize': (float) <microscopy PixelSize>
            * 'fname_prior': An image filename (e.g., .nii.gz) containing processing information
                (e.g., spinal cord segmentation, spinal location or MS lesion classification).
                e.g., spinal cord centerline, used to crop the image prior to segment it if provided.
                The segmentation is not performed on the slices that are empty in this image.
```

Alternatively, is `options` documented anywhere else? I guess my main thinking here is, how does the user discover which options are supported?

Reply:
Oh, 5010993 looks fantastic! Thank you for taking the time to write such thorough documentation. :heart:

Reply:
Very good point! Some of those options go beyond my knowledge of ivadomed at this point, but I tried something.
In 5010993, I took your suggestions and completed the descriptions.

For the 4 postprocessing options, the definitions are a mix of the definitions in ivadomed doc [here](https://ivadomed.org/en/latest/configuration_file.html#postprocessing) and sct_deepseg [here](https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/b97e0a58c47dafef2cb74a07eff1df6681422b94/spinalcordtoolbox/scripts/sct_deepseg.py#L75).
@charleygros, I would need your help to go over the definitions and following concerns/questions:
1. `binarize_prediction`: SCT definition says 0 for soft-seg, isn't it -1?
2. `fill_holes`: is an int (0 or 1) in SCT, so essentially a bool. Should we describe it as a bool or the same as SCT?
3. `keep_largest`: definitions in SCT and ivadomed are not consistent: keep only largest in ivadomed with no argument vs. keep the number specified in SCT?
4. `remove_small`: I was not able to make it work with the current definition.
5. `metadata`: I have no idea how to describe the film metadata, is it the same as `metadata` from [FiLMedUnet params](https://ivadomed.org/en/latest/configuration_file.html#filmedunet)?

> Do we need to keep this description consistent across all of the function docstrings?

I don't think it's necessary, I updated the docstrings other than segment_volume in d3605b5, I think that is enough. Let me know your thoughts.

> Alternatively, is options documented anywhere else? 

I don't think so. But these options are not visible "from outside" except in the arguments from SCT and the future similar implementation with ADS. I think the updated API reference [here](https://ivadomed.org/en/latest/api_ref.html#ivadomed.inference.segment_volume) will be enough.

Reply:
@charleygros, @joshuacwnewton 
I made some more tests and updated the definitions of `options` in `segment_volume`. See updated comment above.
There are some discrepancies between the definition of SCT and ivadomed. It does not affect directly this PR but we may want to address them in the future.
I would be comfortable to merge the PR as is, but I would like to have your approval on the definitions. Thanks!

Reply:
Looks very good to me @mariehbourget !!

Reply:
And I also agree with.. well All your comments regarding discrepancies between the definition of SCT and ivadomed, will need to be addressed in the future yup.

Reply:
Apologies for the delay! 

> For the 4 postprocessing options, the definitions are a mix of the definitions in ivadomed doc [here](https://ivadomed.org/en/latest/configuration_file.html#postprocessing) and sct_deepseg [here](https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/b97e0a58c47dafef2cb74a07eff1df6681422b94/spinalcordtoolbox/scripts/sct_deepseg.py#L75).

Just to make sure I'm understanding correctly, all of the post-processing related code in both projects will ultimately end up at the following places:

* [`inference.py :: set_postprocessing_options()`](https://github.com/ivadomed/ivadomed/blob/e5e3d83d03b0ef9af180448a0ecd72522bed8511/ivadomed/inference.py#L278-L305)
* [`postprocessing.py :: Postprocessing()`](https://github.com/ivadomed/ivadomed/blob/50216fd5a5b573c99a40714280dc84364f1c5e1e/ivadomed/postprocessing.py#L242-L371)

In that case, I definitely agree with your choice to disregard the SCT definitions. It very much looks like SCT will need to update its documentation to properly match what ivadomed is doing. (Especially since `sct_deepseg` just [passes its arguments as-is](https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/b97e0a58c47dafef2cb74a07eff1df6681422b94/spinalcordtoolbox/scripts/sct_deepseg.py#L206) to the ivadomed call.)

Other than that, LGTM! :slightly_smiling_face:

Reply:
> Just to make sure I'm understanding correctly, all of the post-processing related code in both projects will ultimately end up at the following places:
> 
> * [`inference.py :: set_postprocessing_options()`](https://github.com/ivadomed/ivadomed/blob/e5e3d83d03b0ef9af180448a0ecd72522bed8511/ivadomed/inference.py#L278-L305)
> * [`postprocessing.py :: Postprocessing()`](https://github.com/ivadomed/ivadomed/blob/50216fd5a5b573c99a40714280dc84364f1c5e1e/ivadomed/postprocessing.py#L242-L371)

Yes, that is my understanding as well. Also, as far as I know, outside of `ivadomed` only `sct_deepseg` is using `segment_volume` at the moment. Soon it will also be used by ADS for inference with ivadomed trained models (https://github.com/neuropoly/axondeepseg/issues/523).